### ISSUE TITLE
Reducing size and dependencies

### ISSUE THREAD

**Issue by xiam** [getlantern/lantern-android#1]
We're currently generating a huge library that Android developers have to include in their projects, the reason is that we're including a lot of stuff from `flashlight` that might not be used by Android clients.

If we reduce the number of dependencies to only the minimum necessary the size would likely be reduced as well.

I think we could start by trying to get a minimal client based on github.com/getlantern/fronted.

**Comment by xiam** [comment]
Hello @oxtoacart!

I think I need a bit of help here: I tried to import `github.com/getlantern/fronted` from another package, but it fails with an error that is being originated in `fronted` itself, this error I'm talking about is the same we see when trying to get the `fronted` package with `go get -u`:

```
go get -u github.com/getlantern/fronted
# github.com/getlantern/fronted
../fronted/dialer.go:145: unknown enproxy.Conn field 'Addr' in struct literal
../fronted/dialer.go:146: unknown enproxy.Conn field 'Config' in struct literal
../fronted/dialer.go:179: unknown enproxy.Conn field 'Addr' in struct literal
../fronted/dialer.go:180: unknown enproxy.Conn field 'Config' in struct literal
```

At first I thought it was that my `enproxy` was a bit outdated, but the error persist even after updating it:

```
go get -u github.com/getlantern/enproxy
```

Maybe I'm targeting the wrong branches, does the error look familiar to you?

Thanks!

**Comment by oxtoacart** [comment]
@xiam The main repos for a lot of our libraries have probably gotten out of date.  We're doing most of our work in the flashlight-build repo, and only occasionally pushing out to the main repos.

Your best bet is probably to branch the flashlight-build project and then add your work as another item in there.  That'll also help ensure that the mobile and desktop clients are working from the same versions of the libraries.

**Comment by xiam** [comment]
@oxtoacart Yeah, that worked. I'll stick to the flashlight-build repository then. Thank you!

**Comment by oxtoacart** [comment]
@xiam You're welcome!  I hope building using just fronted instead of the full flashlight will make your life easier :)

**Comment by xiam** [comment]
Hey @myleshorton,

I think I'm getting some progress here, I was able to create a working proxy using some Go standard libraries plus `getlantern/balancer` and `getlantern/fronted`. Now I have some questions:
- We're currently using `roundrobin.getiantem.org` as the only fronted server, for testing purposes. I understand that flashlight can update this fronted server list from the network and store it in a configuration file, but in the Go+Android case writing to file system is not that easy, we can surely hack a way to save this list to permanent storage but I think an easier way is providing a method like `AddFrontedServer(addr string)` to the Java program, so the actual Android app is responsible for keeping a list up to date as it's easier for the app to access permanent storage. What do you think?
- What does the `ChainedServers` [property](https://github.com/getlantern/flashlight-build/blob/e430839a775cf67485bc2cb1c9bb63ff90b89adb/src/github.com/getlantern/flashlight/client/balancer.go#L22) do? Would it be useful for Android too?

I'll still need to work a bit on this code, I just wanted to do a quick test to make sure it could be done.

**Comment by myleshorton** [comment]
> We're currently using roundrobin.getiantem.org as the only fronted server, for testing purposes. I understand that flashlight can update this fronted server list from the network and store it in a configuration file, but in the Go+Android case writing to file system is not that easy, we can surely hack a way to save this list to permanent storage but I think an easier way is providing a method like AddFrontedServer(addr string) to the Java program, so the actual Android app is responsible for keeping a list up to date as it's easier for the app to access permanent storage. What do you think?

Yeah interesting. Lantern/Go both just dynamically download them and then store them, which you probably saw, but it could just download, not store, and call the method. Also with built-in automatic updates, it should be less of an issue because we can always just update the hard-coded list with the full client.

The actual domain fronts at this point should be a list of about 1000 domains, and roundrobin.getiantem.org is technically where they front to. So we still want to have a hard-coded list of those domains, associated pinned certs, and IPs hard coded in the client.

**Comment by myleshorton** [comment]
> What does the ChainedServers property do? Would it be useful for Android too?

That basically will relay to a traditional proxy. So flashlight is the first proxy in the "chain" and a normal HTTP proxy running on a server somewhere is the second. I think we should just skip that for Android, though, because we don't have a good blocking-resistant way of distributing them.

**Comment by xiam** [comment]
I compiled a very basic test library with no imports besides the Go-Java bootstrapping libraries, its size was 2732b, so I'm assuming we can't go lower than that. I'm using that number to compute the Relative size column in the following table that illustrates the sizes of the compiled `libgojni.so`:

| Library | Absolute size (b) | Relative size (b) |
| --- | --- | --- |
| Old `libgojni.so` | 8488 | 5756 |
| Current `libgojni.so` | 6452 | 3720 |
| Reduction | 31% | 54% |

I frankly was expecting to see a greater reduction, but cutting a third from the original size is not that bad.

This fronted wrapper has the ability to stop the http server at will, so the proxy can now be stopped and restarted within the Go program space, without the need of using Java threads to avoid blocking! A direct consequence is that the example code was also simplified.

Balancer and reverse proxy are now implemented with channels as in the original flashlight client, I skipped support for ChainedServers.

The issue with the list is still pending, we can have a fixed list and also update that list from the network, as @myleshorton proposed. I'm looking into that too.

**Comment by oxtoacart** [comment]
@xiam Nice!

> I frankly was expecting to see a greater reduction, but cutting a third from the original size is not that bad.

Yeah, that's the one downside to Go.  If you import net and net/http, I bet you'll see that a lot of the size is consumed by those.

The other benefit to what you've done other than size is simplicity, which is worth a lot too.

**Comment by myleshorton** [comment]
> The other benefit to what you've done other than size is simplicity, which is worth a lot too.

Yup I actually wasn't even initially thinking about the size but only the simplicity, so I totally agree there's a huge win there as well. It should also significantly reduce the overall diff. Is this all actually captured in that initial PR, or maybe it's not quite ready to pull yet is that right @xiam? Great stuff either way.

**Comment by xiam** [comment]
@myleshorton I actually closed the initial PR as it is not required anymore :-), no need to add build tags for this one, we just need to keep `balancer` and `fronted` as light as we can. I'm planning to copy `lantern-android` to flashlight-build and create a new PR but I shall finish the issue with the fronted server list first.

**Comment by xiam** [comment]
@oxtoacart should I target the `tcpproxying` branch of `flashlight-build` as the origin for this eventual PR?

**Comment by oxtoacart** [comment]
@xiam You can actually target `devel` now, since we've merged `tcpproxying` into `devel` already.

**Comment by xiam** [comment]
@oxtoacart Another one, I've borrowed the static masquerade and root CA list from [masquerades.go](https://github.com/getlantern/flashlight-build/blob/devel/src/github.com/getlantern/flashlight/config/masquerades.go) and tested it, works beautifully. Now, a huge improvement would be the ability to download and parse the `cloud.yaml` file referred in `./flashlight/genconfig/updateyaml.bash` to keep the internal list up to date (if we can't download this `cloud.yaml` file we can always fallback to the defaults). What's the permantent URL of this file? Thanks again!

**Comment by myleshorton** [comment]
This is the current config @xiam:

https://s3.amazonaws.com/lantern_config/cloud.1.5.11.yaml.gz

**Comment by myleshorton** [comment]
Actually check that. The latest is at:

http://s3.amazonaws.com/lantern_config/cloud.1.6.0.yaml.gz

**Comment by xiam** [comment]
Got it @myleshorton!

**Comment by xiam** [comment]
I solved our pending issue with the configuration file this way: when the application starts it will immediately look for a configuration file over the network, if this can't be done for some reason, or if the configuration file is invalid (wrong format, no useful data, etc) the application will fall back to the default embedded configuration.

I think it would be useful to have a https://getlantern.org/config/cloud.yaml.gz?v=$LANTERN_VERSION URL that redirects to the latest version of this configuration file for a given lantern client version, but for now the S3 URL works beautifully :-).

I'll copy my new repo on the `devel` branch of `flashlight-build` and will update the `README.md` accordingly.

**Comment by myleshorton** [comment]
Awesome @xiam! There are two big issues with the URL you mentioned as is (the https://getlantern.org/config/cloud.yaml.gz?v=$LANTERN_VERSION URL). The first is that getlantern.org is blocked in all of these regions, so we can't run anything on that domain. That's really a larger issue though -- we can't run it on any of our domains or IPs because it'll just get blocked. That's why we only put files like that on either amazon or GitHub over HTTPS (HTTPS since the censor can't tell where you're going then and would have to block all of Amazon or GitHub). In fact, we should really be putting that file on GitHub because S3 is blocked quite a bit. 

That also means, though, that it can't be dynamic either. We can easily just update those fixed URLs for the different versions, though, or do something we just always update like:

http://s3.amazonaws.com/lantern_config/cloud.1.6.0.yaml.gz

**Comment by xiam** [comment]
@myleshorton Oh, of course! I didn't think of that. Maybe we could also consider mirroring the config files on many different locations or CDNs that support HTTPS? The client could try each one of these different sources until succeeding and, if everything else fails, it could always fall back to its default values.

**Comment by myleshorton** [comment]
Yeah totally. At one point we were looking at setting up links on mega but just got pulled away to other things. Right now we just have S3 and GitHub where all _installers_ are automatically copied, but we should really just universally put things like that on any site we have access to that's unblocked.

Oh, also, on that last URL I meant to say:

http://s3.amazonaws.com/lantern_config/latest.yaml.gz 

and not getlantern.org
